core/unit_test/TestComplex.hpp

```diff
@@ -428,4 +428,59 @@ TEST(TEST_CATEGORY, complex_trivially_copyable) {
 #endif
 }
 
+TEST(TEST_CATEGORY, complex_issue_3867) {
+  ASSERT_EQ(Kokkos::pow(Kokkos::complex<double>(2., 1.), 3.),
+            Kokkos::pow(Kokkos::complex<double>(2., 1.), 3));
+  ASSERT_EQ(
+      Kokkos::pow(Kokkos::complex<double>(2., 1.), 3.),
+      Kokkos::pow(Kokkos::complex<double>(2., 1.), Kokkos::complex<double>(3)));
```

Comment:
I omitted 
```diff
diff --git a/core/src/Kokkos_Complex.hpp b/core/src/Kokkos_Complex.hpp
index 623700c2..9f14d762 100644
--- a/core/src/Kokkos_Complex.hpp
+++ b/core/src/Kokkos_Complex.hpp
@@ -773,6 +773,19 @@ struct promote_2 {
 };
 template <class T, class U>
 using promote_2_t = typename promote_2<T, U>::type;
+// complex specializations
+template <class T, class U>
+struct promote_2<Kokkos::complex<T>, U> {
+  using type = Kokkos::complex<promote_2_t<T, U>>;
+};
+template <class T, class U>
+struct promote_2<T, Kokkos::complex<U>> {
+  using type = Kokkos::complex<promote_2_t<T, U>>;
+};
+template <class T, class U>
+struct promote_2<Kokkos::complex<T>, Kokkos::complex<U>> {
+  using type = Kokkos::complex<promote_2_t<T, U>>;
+};
 }  // namespace Impl

 template <class T, class U>
```
on purpose because I thought it was not needed.
I expected the line above would call
https://github.com/kokkos/kokkos/blob/42586296364c85073c44f88c6b2c5db5a61cad8a/core/src/Kokkos_Complex.hpp#L740-L746
instead of
https://github.com/kokkos/kokkos/blob/42586296364c85073c44f88c6b2c5db5a61cad8a/core/src/Kokkos_Complex.hpp#L785-L790
like the error message suggests...

Reply:
This surprised me too.

Reply:
@nliber I think I understand now.

From Daveed'd book (2nd edition):
```
# 15.7.1 Immediate Context
SFINAE protects against attempts to form invalid types or expressions, including errors due to
ambiguities or access control violation that occurs within the *immediate context* of the function 
template substitution. [...] Specifically, during function template substitution for the purpose
of deduction, anything that happens during the instantiation of
* the definition of a class template (i.e., its body and list of base classes),
* [...]
is not part of the immediate context of that function template substitution. [...]
So if substituting the template parameters of a function template declaration requires the
instantiation of the body of a class template because a member of that class is being referred to,
an error during that instantiation is not in the immediate context of the function template
substitution and is therefore a real error (even if another function template matches without
an error).